diff --git a/include/nuttx/video/imgdata.h b/include/nuttx/video/imgdata.h
index 74b8fbef2b2..a7e1ed0010c 100644
--- a/include/nuttx/video/imgdata.h
+++ b/include/nuttx/video/imgdata.h
@@ -58,7 +58,7 @@
 #define IMGDATA_UNINIT(d) \
   ((d)->ops->uninit ? (d)->ops->uninit(d) : -ENOTTY)
 #define IMGDATA_SET_BUF(d, n, f, a, s) \
-  ((d)->ops->set_buf ? (d)->ops->set_buf(d, n, f, a, s) : NULL)
+  ((d)->ops->set_buf ? (d)->ops->set_buf(d, n, f, a, s) : -ENOTTY)
 #define IMGDATA_VALIDATE_FRAME_SETTING(d, n, f, i) \
   ((d)->ops->validate_frame_setting ? \
    (d)->ops->validate_frame_setting(d, n, f, i) : -ENOTTY)
